opm/simulators/linalg/bda/BdaBridge.cpp

```diff
@@ -0,0 +1,242 @@
+/*
+  Copyright 2019 Equinor ASA
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+#include <memory>
+#include <sstream>
+
+#include <opm/common/OpmLog/OpmLog.hpp>
+#include <opm/material/common/Unused.hpp>
+
+#include <opm/simulators/linalg/bda/BdaBridge.hpp>
+#include <opm/simulators/linalg/bda/BdaResult.hpp>
+
+#define PRINT_TIMERS_BRIDGE 0
+
+typedef Dune::InverseOperatorResult InverseOperatorResult;
+
+namespace Opm
+{
+
+BdaBridge::BdaBridge(bool use_gpu_, int linear_solver_verbosity OPM_UNUSED, int maxit OPM_UNUSED, double tolerance OPM_UNUSED) : use_gpu(use_gpu_) {
+#if HAVE_CUDA
+    if(use_gpu){
+    	backend = new cusparseSolverBackend(linear_solver_verbosity, maxit, tolerance);
+    }
+#endif
+}
+
+BdaBridge::~BdaBridge(){
+#if HAVE_CUDA
+	if(use_gpu){
+		delete backend;
+	}
+#endif
+}
+
+#if HAVE_CUDA
+template <class BridgeMatrix>
+int checkZeroDiagonal(BridgeMatrix& mat) {
+	static std::vector<int> diag_indices;   // contains offsets of the diagonal nnzs
+	int numZeros = 0;
+	const int dim = 3;
+	const double zero_replace = 1e-15;
+	double *nnzs = &(mat[0][0][0][0]);
+	if(diag_indices.size() == 0){
+		int N = mat.N()*dim;
+		diag_indices.reserve(N);
+		for(typename BridgeMatrix::const_iterator r = mat.begin(); r != mat.end(); ++r){
+			for(auto c = r->begin(); c != r->end(); ++c){
+				if(r.index() == c.index()){
+					for(int rr = 0; rr < dim; ++rr){
+						// pointer arithmetic
+						int offset = (int)((long unsigned)&(mat[r.index()][c.index()][rr][rr]) - (long unsigned)nnzs); // in bytes
```

Comment:
We also rely on the memory layout in `static_cast<double*>(&(((*mat)[0][0][0][0])))`.
I propose
```
    for(typename BridgeMatrix::size_type i = 0; i < mat.N(); ++i){
    	for(typename BridgeMatrix::size_type j = 0; j < dim; ++j){
    		double val = mat[i][i][j][j];
    		if(val == 0.0){ // could be replaced by '< 1e-30' or similar
    			mat[i][i][j][j] = zero_replace;
    			++numZeros;
    		}
    	}
    }
```
This removes the std::vector diag_indices. mat[i][i] is pretty much guaranteed to exist. For every row, it gets the block on the diagonal (mat[i][i]), and in that block, it checks the values on the diagonal (mat[i][i][j][j]).

Reply:
You kind of have point. But if we rely on something we might want to check that it holds. e.g.
`&((mat.begin()+1).begin()[0][0]) - (&mat.begin().begin()[0][0]) == mat.begin()->size()*blocksize*blocksize;`

Your proposal triggers at least three binary searches per matrix row. I thought the `diag_indices` were there to prevent this.

Reply:
I see. How does your offset work, and how is it used? I can't really follow it.

Reply:
`offset` is the offset of `mat[i][i][j][j]`. `roff` is the offset of the current block row in your cuda matrix values array. `c-r.begin()` is the offset of the nonzero counted from the start of the in block. Then there is a mistake. In the rest please substitute `r` with `rr`:
``offset = roff + (c - r.begin())+ rr * dim + rr;``
The inner loop (over the rows in the matrix block) is omitted. Its loop body is the culculation of `offset`. The last line is in the outer loop.

Reply:
I'm getting the same address for `mat[r.index()][c.index()].begin()` and `mat[r.index()].begin()`, but they are different from `&(mat[r.index()][c.index()][rr][rr])`.

Reply:
Not sure I get where you are trying to go with that statement. Seems like we might be talking about different things.

>
> I'm getting the same address for `mat[r.index()][c.index()].begin()` and `mat[r.index()].begin()`

That can be the	case, but only for the very first non-zero of a	row.
```
if(r.size()>1){
  auto nextC = ++(r.begin());
  assert(&(*mat[r.index()][nextC.index()].begin())!=&(*mat[r.index()].begin()))
}
```
should never abort

> , but they are different from `&(mat[r.index()][c.index()][rr][rr])`.

Only for the case that `rr==0` holds,
`assert(&(mat[r.index()][c.index()][rr][rr])==&(*mat[r.index()][c.index()].begin())`
should not abort.

What I was trying to say is:
No matter what blocksize we have, `c - r.begin()` will tell the consecutive zerostatring index of the current column. Let us say that for the current row, we have a nonzero blocks with column indices `[0, 3, 4, 5]`. If c is positioned at column index `4`, then `c - r.begin()==2` will hold. Does this help.
`c-rbegin

Reply:
The offset (in number of blocks) between the first block of the row and the block that we are interested in (diagonal block) is `int offset = mat[r.index()].find(r.index()).offset())`. This is what you mean by `c - r.begin()`.
I can perform `double *valp = mat[r.index()].getptr() + offset` to get the address of the first value in the block, and then use `valp += dim + 1` to get the 2nd diagonal value of the block.
But it uses some function that are not meant to be used outside of dune-istl.

I think we should only check if the data is contiguous, and use the CPU otherwise. If we have to copy the nonzeroes per row, we'll have 2x44321 `memcpy`s for NORNE. I suspect this defeats the purpose of the GPU.

Reply:
It seems to work using these complicated expressions:
```
for every row:
size_type offset = mat[r.index()].find(r.index()).offset();
for(int rr = 0; rr < dim; ++rr){
    double val = (*(mat[r.index()].getptr() + offset))[rr][rr];
    if(val == 0.0){
        (*(mat[r.index()].getptr() + offset))[rr][rr] = zero_replace;
        ++numZeros;
    }
    diag_indices.emplace_back(offset);
}
```
This stores an offset per bsr row, instead of every csr row. But it is rather ugly.

Reply:
That would do. For beauty, shouldn't this also work:
```
auto diag = r->find(r.index());
assert(diag.index() == r.index());
for(int rr = 0; rr < dim; ++rr){
   auto& val = diag[rr][rr]; // reference to easily change the value
   if (val ==0.0){
      val = zero_replace;
      ++numZeros;
     }
    diag_indices.emplace_back(diag.offset());
}
```
I must say that I do not understand the original offset code any further (even less the more often I look at it). There is another line below these in the original version that does
```
					offset /= sizeof(double);  // convert offset to doubles
```
before storing it. There I am totally lost. Why divide the offset by `8`?

Reply:
For the original offset code:

1. Set `double *nnzs = &(mat[0][0][0][0])`. It points to the start of the underlying array with nonzeroes, which is assumed to be contiguous.
2. For row `r`, and row `rr` in the block, find the address of `mat[r][r][rr][rr]`.
3. Calculate the number of bytes between `nnzs` and the value. Divide by 8 to get the offset in number of doubles.
4. To use the offsets: for every offset, check `nnzs[offset]`. `nnzs` is a `double*` so the index must indicate doubles.

I will try your snippet.

Reply:
I got it working, storing:
```c++
auto diag = r->find(r.index());  // diag is an iterator
assert(diag.index() == r.index());
for(int rr = 0; rr < dim; ++rr){
    auto& val = (*diag)[rr][rr]; // reference to easily change the value
    if (val == 0.0){
        val = zero_replace;
        ++numZeros;
    }
}
diag_indices.emplace_back(diag.offset());
```
Using:
```c++
for(typename BridgeMatrix::iterator r = mat.begin(); r != mat.end(); ++r){
    typename BridgeMatrix::size_type offset = diag_indices[r.index()];
    auto& diag_block = r->getptr()[offset]; // diag_block is a reference to MatrixBlock, located on column r of row r
    for(int rr = 0; rr < dim; ++rr){
        auto& val = diag_block[rr][rr];
        if(val == 0.0){
            val = zero_replace;
            ++numZeros;
        }
    }
}
```
But it still uses the `r->getptr()` construct. I think the only alternative is to use `find()`, which performs the binary search.

Reply:
I would have assumed that `r->begin()+offset` does the same as `r->getptr()[offset]`. Anyway, both look good and it seems a matter of test. You do not need to invest more time here. Would you please push your changes (including the replacement of tabs with space and give me a heads up. I guess this is good to be merged.